File: backend/app/transpose.py
"""
Transposição de cifras no padrão brasileiro: acordes marcados entre colchetes
dentro do texto da letra, ex: "[C]Vem, ó [G]Espírito de [Am/E]Deus".

Suporta: nota base (A-G), sustenido (#) ou bemol (b), sufixos (m, 7, maj7, sus4,
dim, aug, 9, 11, 13, add9 ...) e baixo/inversão após "/".
"""
import re

NOTES_SHARP = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"]
FLAT_TO_SHARP = {
    "Db": "C#", "Eb": "D#", "Gb": "F#", "Ab": "G#", "Bb": "A#",
    "Cb": "B", "Fb": "E",
}

# Captura: nota raiz [A-G](#|b)? + resto do acorde (sufixo) + baixo opcional (/nota)
CHORD_RE = re.compile(
    r"\[([A-G](?:#|b)?)([^/\]]*)(?:/([A-G](?:#|b)?))?\]"
)


def _normalize(note: str) -> str:
    return FLAT_TO_SHARP.get(note, note)


def _shift(note: str, semitons: int) -> str:
    note = _normalize(note)
    idx = NOTES_SHARP.index(note)
    new_idx = (idx + semitons) % 12
    return NOTES_SHARP[new_idx]


def transpose_chord(match: re.Match, semitons: int) -> str:
    root, suffix, bass = match.group(1), match.group(2), match.group(3)
    new_root = _shift(root, semitons)
    result = f"[{new_root}{suffix}"
    if bass:
        result += f"/{_shift(bass, semitons)}"
    result += "]"
    return result


def transpose_cifra(cifra: str, semitons: int) -> str:
    if semitons % 12 == 0:
        return cifra
    return CHORD_RE.sub(lambda m: transpose_chord(m, semitons), cifra)


def transpose_key(tom: str, semitons: int) -> str:
    """Transpõe apenas o nome de um tom, ex: 'D' + 2 -> 'E'."""
    return _shift(tom, semitons)

File: frontend/src/lib/transpose.ts
// Transposição de cifras — mesma lógica do backend (app/transpose.py),
// reimplementada em TS para funcionar 100% offline no cliente.

const NOTES_SHARP = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"];
const FLAT_TO_SHARP: Record<string, string> = {
  Db: "C#", Eb: "D#", Gb: "F#", Ab: "G#", Bb: "A#", Cb: "B", Fb: "E",
};

const CHORD_RE = /\[([A-G](?:#|b)?)([^/\]]*)(?:\/([A-G](?:#|b)?))?\]/g;

function normalize(note: string): string {
  return FLAT_TO_SHARP[note] ?? note;
}

function shift(note: string, semitons: number): string {
  const norm = normalize(note);
  const idx = NOTES_SHARP.indexOf(norm);
  const newIdx = ((idx + semitons) % 12 + 12) % 12;
  return NOTES_SHARP[newIdx];
}

export function transposeCifra(cifra: string, semitons: number): string {
  if (semitons % 12 === 0) return cifra;
  return cifra.replace(CHORD_RE, (_match, root: string, suffix: string, bass?: string) => {
    const newRoot = shift(root, semitons);
    let result = `[${newRoot}${suffix}`;
    if (bass) result += `/${shift(bass, semitons)}`;
    result += "]";
    return result;
  });
}

export function transposeKey(tom: string, semitons: number): string {
  return shift(tom, semitons);
}

/** Renderiza a cifra em texto (com [Acordes]) como blocos {chord, lyric} para exibição. */
export function parseCifraLine(line: string): { chord: string | null; lyric: string }[] {
  const parts: { chord: string | null; lyric: string }[] = [];
  let lastIndex = 0;
  let currentChord: string | null = null;
  const re = /\[([^\]]+)\]/g;
  let match: RegExpExecArray | null;
  while ((match = re.exec(line)) !== null) {
    const textBefore = line.slice(lastIndex, match.index);
    if (textBefore) parts.push({ chord: currentChord, lyric: textBefore });
    currentChord = match[1];
    lastIndex = re.lastIndex;
  }
  const rest = line.slice(lastIndex);
  parts.push({ chord: currentChord, lyric: rest });
  return parts;
}
